pilot/pkg/config/kube/gateway/inferencepool_collection.go

```diff
@@ -537,33 +539,60 @@ func (c *Controller) reconcileShadowService(
 		existingService := ptr.Flatten(servicesCollection.GetKey(pool.shadowService.key.String()))
 
 		// Check if we can manage this service
-		var existingLabels map[string]string
 		if existingService != nil {
-			existingLabels = existingService.GetLabels()
 			canManage, _ := c.canManageShadowServiceForInference(existingService)
 			if !canManage {
 				log.Debugf("skipping service %s/%s, already managed by another controller", key.Namespace, key.Name)
 				return nil
 			}
 		}
 
-		service := translateShadowServiceToService(existingLabels, pool.shadowService, pool.extRef)
+		service := translateShadowServiceToService(pool.shadowService, pool.extRef)
 
 		var err error
 		if existingService == nil {
 			// Create the service if it doesn't exist
 			_, err = svcClient.Create(service)
 		} else {
-			// TODO: Don't overwrite resources: https://github.com/istio/istio/issues/56667
-			service.ResourceVersion = existingService.ResourceVersion
-			_, err = svcClient.Update(service)
+			// Patch the service if it already exists
+			var patchBytes []byte
+			patchBytes, err = c.generateShadowServicePatch(existingService, service)
+			if err != nil {
+				return err
+			}
+			_, err = svcClient.Patch(existingService.Name, existingService.Namespace, types.MergePatchType, patchBytes)
 		}
 
 		return err
 	}
 }
 
-// canManage checks if a service should be managed by this controller
+func (c *Controller) generateShadowServicePatch(existing, desired *corev1.Service) ([]byte, error) {
+	updated := existing.DeepCopy()
+
+	// Merge labels and annotations
+	updated.Annotations = maps.MergeCopy(existing.Annotations, desired.Annotations)
```

Comment:
If we are doing a patch won't k8s do the merge for us?

Also this means we can never remove things IIUC patches correctly (not certain I do)

Reply:
Yeah, but the thing is that to generate a good patch, you need full information about what the current resource status is and what it should look like at the end, that's why the merges are required. But your point about SSA is completely correct, I'll look into that, I guess using Patch is bit old-school now

Reply:
Actually it might not even be required here I realize; I was worried jsonmerge would generate a `replace` directive for the entire annotations/labels if I don't do this. Not sure if that's the case though. Anyway, it's not required for SSA so we can drop it.